smart-contracts/wasm-chain-integration/src/utils.rs

```diff
@@ -273,34 +279,30 @@ impl<'a, R: RngCore, BackingStore: trie::BackingStoreLoad> machine::Host<Artifac
                 self.slot_time = Some(slot_time);
             }
             "get_slot_time" => {
-                let slot_time = self.slot_time.context(unset_err("slot_time"))?;
+                let slot_time = self.slot_time.ok_or(CallErr::Unset("slot_time"))?;
                 stack.push_value(slot_time);
             }
             "set_receive_self_address" => {
                 let addr_ptr = unsafe { stack.pop_u32() };
 
                 let mut cursor = Cursor::new(memory);
-                cursor.seek(SeekFrom::Start(addr_ptr)).map_err(|_| anyhow!(seek_err))?;
+                cursor.seek(SeekFrom::Start(addr_ptr)).map_err(|_| CallErr::Seek)?;
 
                 self.address = Some(ContractAddress::deserial(&mut cursor)?);
             }
             "get_receive_self_address" => {
-                let addr_ptr = unsafe { stack.pop_u32() };
-                let mut cursor = Cursor::new(memory);
-
-                cursor.seek(SeekFrom::Start(addr_ptr)).map_err(|_| anyhow!(seek_err))?;
-
+                let addr_ptr = unsafe { stack.pop_u32() } as usize;
                 self.address
-                    .context(unset_err("slot_time"))?
-                    .serial(&mut cursor)
+                    .ok_or(CallErr::Unset("address"))?
+                    .serial(&mut &mut memory[addr_ptr..])
                     .map_err(|_| anyhow!("Unable to serialize the self address"))?;
```

Comment:
```suggestion
                    .map_err(|_| CallErr::Write)?;
```

Reply:
Hmm, I didn't like this as it removes context, but it's of course what I've already done elsewhere. I added a "CallErr::Serial" to the error enum that takes in a `'static &str` so:

`CallErr::Serial("self_address")`